cirq/circuits/qasm_output.py

```diff
@@ -249,52 +249,43 @@ def _write_operations(self,
                           output: Callable[[str], None],
                           output_line_gap: Callable[[int], None],
                           top=True) -> None:
-        for op in ops.flatten_op_tree(op_tree):
-            out = protocols.qasm(op, args=self.args, default=None)
-            if out is not None:
-                output(out)
-                continue
-
-            if isinstance(op, ops.GateOperation):
-                comment = 'Gate: {!s}'.format(op.gate)
-            else:
-                comment = 'Operation: {!s}'.format(op)
-
-            decomp = protocols.decompose_once(op, None)
-            if decomp is not None:
-                if top:
-                    output_line_gap(1)
-                    output('// {}\n'.format(comment))
-                self._write_operations(decomp,
-                                       output,
-                                       output_line_gap,
-                                       top=False)
-                if top:
-                    output_line_gap(1)
-                continue
-
-            mat = protocols.unitary(op, None) if len(op.qubits) <= 2 else None
-            if mat is not None and len(op.qubits) == 1:
-                u_op = QasmUGate.from_matrix(mat).on(*op.qubits)
-                if top:
-                    output_line_gap(1)
-                    output('// {}\n'.format(comment))
-                output(cast(str, protocols.qasm(u_op, args=self.args)))
-                if top:
-                    output_line_gap(1)
-                continue
-
-            if mat is not None and len(op.qubits) == 2:
-                u_op = QasmTwoQubitGate.from_matrix(mat).on(*op.qubits)
-                if top:
-                    output_line_gap(1)
-                    output('// {}\n'.format(comment))
-                self._write_operations((u_op,),
-                                       output,
-                                       output_line_gap,
-                                       top=False)
-                if top:
-                    output_line_gap(1)
-                continue
-
-            raise ValueError('Cannot output operation as QASM: {!r}'.format(op))
+        for op_original in ops.flatten_op_tree(op_tree):
+            keep = lambda _op: (protocols.qasm(_op, args=self.args,
+                                               default=None)
+                                is not None)
+            for op in protocols.decompose(op_original, keep=keep, on_stuck_raise=None):
```

Comment:
Okay, basically, the issue you're running into is that the code is losing track of whether or not the initial operations are being decomposed or not and so the logic around commenting is going wrong.

To fix the issue you need to move this `out = qasm(...) if out: output(out); continue` block to the start of the outer loop, just before the current assignment of `keep`.

Reply:
Moving that block up didn't make it any better. Getting slightly better results with setting comment_matrix_fallbacks to op==op_original (or just setting to true only if it's the first element in the inside loop). But I'm sure these are not the correct solutions.
https://pastebin.com/Fc6X74ii is the new diff